File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int ind1=text1.length()-1;
        int ind2=text2.length()-1;
        int[][] dp=new int[ind1+1][ind2+1];
        for(int i=0;i<dp.length;i++){
            Arrays.fill(dp[i],-1);
        }
        return solve(ind1,ind2,text1,text2,dp);
    }
    public int solve(int ind1,int ind2,String text1,String text2,int[][] dp){
        if(ind1<0 ||ind2<0){
            return 0;
        }
        if(dp[ind1][ind2]!=-1){
            return dp[ind1][ind2];
        }
        if(text1.charAt(ind1)==text2.charAt(ind2)){
            return dp[ind1][ind2]=1+solve(ind1-1,ind2-1,text1,text2,dp);
        }
        return dp[ind1][ind2]=0+Math.max(solve(ind1-1,ind2,text1,text2,dp),solve(ind1,ind2-1,text1,text2,dp));
    }
}

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
    int solve(int i ,int j,string &text1,string &text2)
    {
        if(i==0 && j==0) return text1[i]==text2[j];
        if(i==0 && j>0)
        {
            if(text1[i]==text2[j]) return 1;
            return solve(i,j-1,text1,text2);
        }
        if(i>0 && j==0)
        {
            if(text1[i]==text2[j]) return 1;
            return solve(i-1,j,text1,text2);
        }
        
        if(text1[i]==text2[j]) return 1+solve(i-1,j-1,text1,text2);

        return max(solve(i-1,j,text1,text2),solve(i,j-1,text1,text2));
    }
public:
    int longestCommonSubsequence(string text1, string text2) {
        int n=text1.size();
        int m=text2.size();
        
        vector<vector<int>> dp(n+1,vector<int>(m+1,0));
        // if(text1[0]==text2[0]) dp[0][0]=1;
        // for(int j=0;j<m;j++)
        // {
        //      if(text1[0]==text2[j]) dp[0][j]=1;
        // }
        // for(int i=0;i<n;i++)
        // {
        //      if(text1[i]==text2[0]) dp[i][0]=1;
        // }
        for(int i=1;i<=n;i++)
        {
            for(int j=1;j<=m;j++)
            {
                if(text1[i-1]==text2[j-1]) dp[i][j]=1+dp[i-1][j-1];

               else  dp[i][j]= max(dp[i-1][j],dp[i][j-1]);
            }
        }
        return dp[n][m];
    }
};